### ISSUE TITLE
Can we use actions with directional parameters in tables

### ISSUE THREAD

**Issue by vgurevich** [p4lang/p4c#2501]
Attached is s trivial program 
[duplicate_action_name.p4.txt](https://github.com/p4lang/p4c/files/5088529/duplicate_action_name.p4.txt)
that declares an action with a directional parameter and then tries to use it in a table multiple times, each being bound to a different argument. The compilation fails with the following message:

```
duplicate_action_names.p4(23): [--Werror=duplicate] error: dec: Duplicates declaration dec
            dec(meta.b);
            ^^^^^^^^^^^
duplicate_action_names.p4(21)
            dec(meta.a);
            ^^^^^^^^^^^
duplicate_action_names.p4(25): [--Werror=duplicate] error: dec: Duplicates declaration dec
            dec(meta.c);
            ^^^^^^^^^^^
duplicate_action_names.p4(21)
            dec(meta.a);
            ^^^^^^^^^^^
error: 2 errors encountered, aborting compilation
```

This seems to contradict the intention of the example, provided at the end of the section 13.2.1.2 of the spec.

I would also mention that the spec is a little quiet about what might happen in this particular case, but I tried to play it nice and specifically take case of naming the actions differently by adding `@name()` annotation. I do not think it should be required (although it is not clear how should the compiler rename these by default), but in any case it didn't help.

**Comment by jnfoster** [comment]
Section 13.2.1.2 says

> Each action in the list of actions for a table must have a distinct name

For the purposes of satisfying this stipulation, I don't believe `@name` annotations are meant to modify the name of the action. Hence, the program is correctly rejected by `p4c`.

**Comment by vgurevich** [comment]
@jnfoster -- in this case we should probably tighten the spec and update it accordingly by explicitly discussing this case.

However, this makes actions with directional parameters a little less useful -- essentially one would have to define wrapper actions first before using them in tables -- not a biggy, but even more lines of code. 

I am also not sure, (given the fact that these actions cannot be used without being bounded to specific parameter values) whether the bounded instances of these actions do necessarily have the same name. Again, can go either way, but then we need to clarify in the spec.

For example, consider the following code:
```
table t1 {
    actions = { dec(meta.a) }
}

table t2 {
    actions = { dec(meta.b) }
}
```
The action "dec" will then have quite different semantics between tables and should then probably officially be named as `.Ingress.t1.dec` or `Ingress.t2.dec` otherwise it will be quite difficult to even talk about the program.

**Comment by jnfoster** [comment]
Responses: 
- I agree that clarifying the restrictions on names in the same table would be good. We do have an example with `a` and `.a` but we should also point out that partially applying the same action to different data plane arguments does not yield a differently named action. 

- Likewise, we should clarify that `@name` annotations don't affect the names of actions, either for the purposes of disambiguating or for the control-plane. (Or change the implementation so they do.)

- I think the section on [control-plane names](https://p4.org/p4-spec/docs/P4-16-v1.2.1.html#sec-cp-names) already says the two partial applications of `dec` have the qualified names you wrote down. So I think we're good there.

**Comment by vgurevich** [comment]
@jnfoster -- here are a couple of thoughts:

1. Frankly speaking, the current example with `a` and `.a` in the spec is not very convincing: I might be wrong, but `a` (which represents `c.a` in that context) and `.a` look like two pretty different names to me. Either we need to change the example or we need to define what does it mean for two names to be the same
2. For the purposes of this issue, I think that we should at the very minimum allow `@name()` annotations in that context. What is the reason not to? Why can one rename keys, but not actions? Ideally, I think we should have first class syntax to rename both -- that will help if we want to use these renamed actions in `const entries`
3. Your last statement (that actions have the same scope as everything else) seems to confirm my earlier point that `a` and `.a` are two different names. It doesn't help my cause, though :(.

Having said that, I think that the proposed approach limits the usefulness of the construct. Actions with directional parameters are specifically designed for code reuse and thus there should be a way to instantiate and use them in many different places without restrictions. Otherwise they will remain an obscure feature and people will be forced to use preprocessor macros instead.

Here is how I'd see this done ideally:
```
table t {
    keys = { 
        ip_valid   ::= hdr.ipv4.isValid() || hdr.ipv6.isValid() : exact;
        mpls_valid ::= hdr.mpls.isValid()                       : exact; 
    actions {
        dec_a ::= dec(meta.a);
        dec_b ::= dec(meta.b);
        dec_c ::= dec(meta.c);
   }
   const entries {
       (true, true)  : dec_a();
       (true, false) : dec_b();
       ...
   }
}
```

**Comment by mbudiu-vmw** [comment]
Yes, a, and .a are different names. In the future, when we have modules, we will have more general paths to prefix names; .a is just a simple path.

`@name` annotation do not rename actions, they just specify a control-plane name for an object. How would you write a `switch` statement if you can have duplicate action names?

**Comment by vgurevich** [comment]
@mbudiu-vmw -- thanks. So, it looks like we have a simple issue with `a` and `.a` both in the spec and in the compiler, although most probably it is fairly theoretical.

I see your point about the connection between the action names and `actionRun()`. If the only place that `@name` affects is the control plane and we can't use this assigned name further in P4 (and I understand that), then we should seriously consider the construct similar to what I proposed in [one the the comments above](https://github.com/p4lang/p4c/issues/2501#issuecomment-677099188). What do you think?

**Comment by mbudiu-vmw** [comment]
Yes, I see no problem having a compact syntax to "rename" actions. Do you want to write a PR against the spec?
I will put this in the LDWG wiki.

**Comment by smolkaj** [comment]
> ```
> table t {
>     keys = { 
>         ip_valid   ::= hdr.ipv4.isValid() || hdr.ipv6.isValid() : exact;
>         mpls_valid ::= hdr.mpls.isValid()                       : exact; 
>     actions {
>         dec_a ::= dec(meta.a);
>         dec_b ::= dec(meta.b);
>         dec_c ::= dec(meta.c);
>    }
>    const entries {
>        (true, true)  : dec_a();
>        (true, false) : dec_b();
>        ...
>    }
> }
> ```

@vgurevich, just a quick thought regarding syntax. Given that `keys`, `actions`, and `const entries` are all conceptually dictionaries, I feel it would be nice to use the same syntax to specify all three mappings (though I recognize this may not be possible in a backwards compatible way, since `:` is also used to specify the match kind).

**Comment by mbudiu-vmw** [comment]
Borrowing from the safe unions proposal an alternative would be 
`a(x) as ax`

**Comment by mbudiu-vmw** [comment]
`dec_a = dec(meta.a)`

**Comment by mbudiu-vmw** [comment]
We need a common solution for actions, keys, package arguments, and union labels.

**Comment by mbudiu-vmw** [comment]
t@expr
